src/ui/public/vis_maps/maps_renderbot.js

```diff
@@ -1,86 +1,274 @@
+      });
+      this._kibanaMap.on('zoomend', () => {
 
-    if (this.mapsData) {
-      this.mapsVis.render(this.mapsData, this.uiState);
+        const isAutoPrecision = _.get(this._chartData, 'geohashGridAgg.params.autoPrecision', true);
+        if (!isAutoPrecision) {
+          return;
+        }
+
+        this._dataDirty = true;
+        if (precisionChange) {
+          courier.fetch();
+        } else {
+          this._recreateGeohashLayer();
+          this._dataDirty = false;
+          this._doRenderComplete();
+        }
+      });
+
+
+      this._kibanaMap.on('drawCreated:rectangle', event => {
+        addSpatialFilter(_.get(this._chartData, 'geohashGridAgg'), 'geo_bounding_box', event.bounds);
+      });
+      this._kibanaMap.on('baseLayer:loaded', () => {
+        this._baseLayerDirty = false;
+        this._doRenderComplete();
+      });
+      this._kibanaMap.on('baseLayer:loading', () => {
+        this._baseLayerDirty = true;
+      });
+    }
+
+    _recreateGeohashLayer() {
+      if (this._geohashLayer) {
+        this._kibanaMap.removeLayer(this._geohashLayer);
+      }
+      if (!this._geohashGeoJson) {
+        return;
+      }
+      const geohashOptions = this._getGeohashOptions();
+      this._geohashLayer = new GeohashLayer(this._chartData.geoJson, geohashOptions, this._kibanaMap.getZoomLevel());
+      this._kibanaMap.addLayer(this._geohashLayer);
     }
-  };
 
-  MapsRenderbot.prototype._getMapsParams = function () {
-    const self = this;
 
-    return _.assign(
-      {},
-      self.vis.type.params.defaults,
-      {
-        type: self.vis.type.name,
-        // Add attribute which determines whether an index is time based or not.
-        hasTimeField: self.vis.indexPattern && self.vis.indexPattern.hasTimeField()
-      },
-      self.vis.params
-    );
-  };
+    /**
+     * called on data change
+     * @param esResponse
+     */
+    render(esResponse) {
+      this._dataDirty = true;
+      this._kibanaMapReady.then(() => {
+        this._chartData = this._buildChartData(esResponse);
+        this._geohashGeoJson = this._chartData.geoJson;
+        this._recreateGeohashLayer();
+        this._useUIState();
+        this._kibanaMap.resize();
+        this._dataDirty = false;
+        this._doRenderComplete();
+      });
+    }
 
-  MapsRenderbot.prototype.buildChartData = buildChartData;
-  MapsRenderbot.prototype.render = function (esResponse) {
-    this.mapsData = this.buildChartData(esResponse);
-    return AngularPromise.delay(1).then(() => {
-      this.mapsVis.render(this.mapsData, this.uiState);
-    });
-  };
+    destroy() {
+      if (this._kibanaMap) {
+        this._kibanaMap.destroy();
+      }
+    }
 
-  MapsRenderbot.prototype.destroy = function () {
-    const self = this;
+    /**
+     * called on options change (vis.params change)
+     */
+    updateParams() {
 
-    const mapsVis = self.mapsVis;
+      this._paramsDirty = true;
+      this._kibanaMapReady.then(_ => {
```

Comment:
++ We shouldn't include variables that we don't use.  And the only thing that should be using `_` as a variable is lodash, just for clarity's sake.